src/decode/ykpt/packet_parser/mod.rs

```diff
@@ -0,0 +1,303 @@
+    /// values are sometimes compressed using bits from the previous TIP value.
+    prev_tip: usize,
+}
+
+/// Attempt to read the packet of type `$packet` using deku. On success wrap the packet up into the
+/// corresponding discriminant of `Packet`.
+macro_rules! read_to_packet {
+    ($packet: ty, $bits: expr, $discr: expr) => {
+        <$packet>::read($bits, ()).and_then(|(r, p)| Ok((r, $discr(p))))
+    };
+}
+
+/// Same as `read_to_packet!`, but with extra logic for dealing with packets which encode a TIP.
+macro_rules! read_to_packet_tip {
+    ($packet: ty, $bits: expr, $discr: expr, $prev_tip: expr) => {
+        <$packet>::read($bits, ()).and_then(|(r, p)| {
+            let ret = if p.needs_prev_tip() {
+                Ok((r, $discr(p, Some($prev_tip))))
+            } else {
+                Ok((r, $discr(p, None)))
+            };
+            ret
+        })
+    };
+}
+
+impl<'t> PacketParser<'t> {
+    pub(super) fn new(bytes: &'t [u8]) -> Self {
+        Self {
+            bytes,
+            state: PacketParserState::Init,
+            prev_tip: 0,
+        }
+    }
+
+    fn mk_parse_error(&mut self) -> HWTracerError {
+        let mut s = String::new();
+        write!(
+            &mut s,
+            "In state '{:?}', failed to parse packet: ",
+            self.state
+        )
+        .unwrap();
+        write!(&mut s, "{}", self.byte_stream_str(8, ", ")).unwrap();
+        HWTracerError::TraceParseError(s)
+    }
+
+    /// Attempt to parse a packet of the specified `PacketKind`.
+    fn parse_kind(&mut self, kind: PacketKind) -> Option<Packet> {
+        let bits = BitSlice::from_slice(self.bytes).ok()?;
+        let parse_res = match kind {
+            PacketKind::PSB => {
+                read_to_packet!(PSBPacket, bits, Packet::PSB)
+            }
+            PacketKind::CBR => read_to_packet!(CBRPacket, bits, Packet::CBR),
+            PacketKind::PSBEND => read_to_packet!(PSBENDPacket, bits, Packet::PSBEND),
+            PacketKind::PAD => read_to_packet!(PADPacket, bits, Packet::PAD),
+            PacketKind::MODE => read_to_packet!(MODEPacket, bits, Packet::MODE),
+            PacketKind::TIPPGE => {
+                read_to_packet_tip!(TIPPGEPacket, bits, Packet::TIPPGE, self.prev_tip)
+            }
+            PacketKind::TIPPGD => {
+                read_to_packet_tip!(TIPPGDPacket, bits, Packet::TIPPGD, self.prev_tip)
+            }
+            PacketKind::ShortTNT => read_to_packet!(ShortTNTPacket, bits, Packet::ShortTNT),
+            PacketKind::LongTNT => read_to_packet!(LongTNTPacket, bits, Packet::LongTNT),
+            PacketKind::TIP => read_to_packet_tip!(TIPPacket, bits, Packet::TIP, self.prev_tip),
+            PacketKind::FUP => read_to_packet_tip!(FUPPacket, bits, Packet::FUP, self.prev_tip),
+            PacketKind::CYC => read_to_packet!(CYCPacket, bits, Packet::CYC),
+        };
+        if let Ok((remain, pkt)) = parse_res {
```

Comment:
This is probably shorter as a `match`.

Reply:
It turns out 7 lines instead of 6:

```
        if let Ok((remain, pkt)) = parse_res {
            self.bytes = remain.as_raw_slice();
            Some(pkt)
        } else {
            None
        }
```

vs. 

```
        match parse_res {
            OK((remain, pkt)) => {
                self.bytes = remain.as_raw_slice();
                Some(pkt)
            },
            Err(_) => None,
        }
```

Reply:
OK.